runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -7607,13 +7626,16 @@ reservationLockExit(TR::Node *node, int32_t lwOffset, TR::Register *objectClassR
       generateRIInstruction  (cg, addImmOp, node, monitorReg,
          (uintptr_t) (isPrimitive ? LOCK_INC_DEC_VALUE : -LOCK_INC_DEC_VALUE) & 0x0000FFFF);
       generateRXInstruction(cg, storeOp, node, monitorReg, generateS390MemoryReference(objReg, lwOffset, cg));
-
+#if defined(TR_TARGET_64BIT) && (JAVA_SPEC_VERSION >= 19)
+      generateS390LabelInstruction(cg, TR::InstOpCode::label, node, primitiveBranchLabel);
+      generateSIInstruction(cg, TR::InstOpCode::AGSI, node, generateS390MemoryReference(cg->getMethodMetaDataRealRegister(), fej9->thisThreadGetOwnedMonitorCountOffset(), cg), -1);
+#else
```

Comment:
Did we change the logic in the `isPrimitive` equal to `true` case here?  Assuming the masked AND operation against `OBJECT_HEADER_LOCK_RECURSION_MASK` returns COND_EQ (result = 0)... the Java < 19 logic will fall through to the callLabel.  In the Java 19+ logic, we will unconditionally go to `helperReturnOOLLabel` instead.

I'm not even 100% sure if we should be decrementing the `OwnedMonitorCountOffset` in this `isPrimitive` case where we take the fall-through path on the cond branch after AND instruction above.

Reply:
I see the flaw in the code, fixing it now. Though, I have a doubt with this in general primitive lock reservation code. In the code, if the lockward do not have Lock recursion mask set in primitive case, it will fall through which will increment the lockward, store it back to object and call helper [1]. 
Looking at the reservationLockEnter code, we only increment the counter in the case of non primitive. Do not do anything in case of primitive. Trying to find a documentation on lock reservation JIT implementation as well as other code gen code to see what is the correct thing to do. 

[1]. https://github.com/eclipse-openj9/openj9/blob/a96598460eab6fc6fdb2fdf8208f77fc6df2790f/runtime/compiler/z/codegen/J9TreeEvaluator.cpp#L7597-L7612